### PR TITLE
Add Authors to the list of things you can search for in the Object Selection menu

### PR DESCRIPTION
**Pull request by karst**
Thanks SpaceK for providing their code. My own way of doing it was pretty similar but i kept running into one issue which is fixed here.

This feature/improvement will definitely help with the object creation community. I really hope it prompts people to stop putting their names in the object names.

I am gonna see if I can add this author trivia to some other windows so people get credit where it's due.

### CODE REVIEW

**Comment by spacek531** [comment]
Adding the authors field in e.g. scenery window should perhaps be another PR?

**Comment by karst** [comment]
Yea, they should definitely be separated PRs, no worries

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 1]
```diff
@@ -1343,7 +1343,23 @@ class EditorObjectSelectionWindow final : public Window
...
+        }
+
+        // Check in the searched string exists in the authors field
+        bool inAuthor = false;
+        for (auto author : item->Authors)
+        {
+            inAuthor = String::ToUpper(author).find(filterUpper) != std::string::npos;
+            if (inAuthor)
+            {
+                break;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:1358]: just return true here no need to keep searching.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 2]
```diff
@@ -1343,7 +1343,23 @@ class EditorObjectSelectionWindow final : public Window
...
+            }
+        }
+
+        return inAuthor;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:1362]: then you can return false here.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 3]
```diff
@@ -1343,7 +1343,23 @@ class EditorObjectSelectionWindow final : public Window
...
     }
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:1363]: You might as well just create a short helper function whilst you are at it.
```
static bool isInAuthor(const std::vector<string>& authors){
    for (auto& author : authors)
        {
            inAuthor = String::ToUpper(author).find(filterUpper) != std::string::npos;
            if (inAuthor)
            {
                return true;
            }
    }
    return false;
}
```

Comment (IntelOrca) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:1363]: Can we also create a helper function to find a case-insensitive string? `String::ToUpper` is expensive, and you care calling it for every item in the list.

Comment (karst) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:1363]:
```C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,50): error C2065:
 'string': undeclared identifier [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,43): error C2923:
 'std::vector': 'string' is not a valid template type argument for parameter '_Ty' [C:\Users\Gebruiker\Documents\GitHub
\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,43): error C2976:
 'std::vector': too few template arguments [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.
vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,1): error C2955:
'std::vector': use of class template requires template argument list [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\
openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,59): error C2955:
 'std::vector': use of class template requires template argument list [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src
\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,21): error C2267:
 'isInAuthor': static functions with block scope are illegal [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2
-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1352,21): error C2601:
 'isInAuthor': local function definitions are illegal [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\lib
openrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1354,33): error C2663:
 'std::vector<_Ty,_Alloc>::_Unchecked_begin': 2 overloads have no legal conversion for 'this' pointer [C:\Users\Gebruik
er\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1354,33): error C2663:
 'std::vector<_Ty,_Alloc>::_Unchecked_end': 2 overloads have no legal conversion for 'this' pointer [C:\Users\Gebruiker
\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1355,1): error C3536:
'<begin>$L0': cannot be used before it is initialized [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\lib
openrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1355,1): error C3536:
'<end>$L0': cannot be used before it is initialized [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libop
enrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1354,40): error C2100:
 illegal indirection [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1356,17): error C2065:
 'inAuthor': undeclared identifier [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1356,1): error C2664:
'std::string String::ToUpper(std::string_view)': cannot convert argument 1 from 'int' to 'std::string_view' [C:\Users\G
ebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1356,57): error C2065:
 'filterUpper': undeclared identifier [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxpr
oj]
  C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\windows\EditorObjectSelection.cpp(1357,21): error C2065:
 'inAuthor': undeclared identifier [C:\Users\Gebruiker\Documents\GitHub\OpenRCT2\src\openrct2-ui\libopenrct2ui.vcxproj]

    0 Warning(s)
    16 Error(s)
```

This is basically the same issue I kept running into before by applying the same logic as used for previous commands.

> Can we also create a helper function to find a case-insensitive string? 

I am just repeating what is done in previous lines. I don't quite know how I'd go about creating a helper function, unless you mean that one of you guys change this afterwards at some point?

Comment (duncanspumpkin) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:1361]: I've applied the fix myself.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**